### PR TITLE
Fix #26

### LINKED ISSUE #26: constructor gets confused by function pointers
https://github.com/atom/language-c/issues/129

```
struct foo {
    int (*bar)();
};

union baz {
    int (*quz)();
};
```

int (*bar) is tagged as a constructor

### PR DESCRIPTION
**Pull request by matter123**
This is the Fix for #26 with updated test extracted from #135.

### CODE REVIEW

**Comment by jeff-hykin** [comment]
Screen Shot 2019-05-10 at 5 45 43 PM

**Comment by jeff-hykin** [comment]
I'm not sure why the `}` is missing highlighting

**Comment by matter123** [comment]
the a isn't valid c++ code

**Comment by matter123** [comment]
https://godbolt.org/z/Jwinbc

**Comment by jeff-hykin** [comment]
For the brackets at the end of the function pointer, would that only be for declaring an array of function pointers?

**Comment by matter123** [comment]
No its actually for declaring the the return type is an array https://cdecl.org/?q=int+%28*foo%29%28%29%5B%5D

**Comment by jeff-hykin** [comment]
haha that is a great link

**Comment by matter123** [comment]
Oh sorry the current_array brackets yes declare an array of function pointers.

**Comment by jeff-hykin** [comment]
Do you understand this? `int (*(*foo)(void ))[3]`
https://cdecl.org/?q=int+%28*foo%29%28%29%5B%5D
I'm still not sure what exactly is being declared. I mostly dont understand the `[3]`

**Comment by matter123** [comment]
So it helps to remember that its the same as `int *(*foo)(void )[3]` which is a pointer to a function take takes nothing and returns a pointer to an array of 3 integers.  which is not actually valid https://godbolt.org/z/8sxwSh

**Comment by matter123** [comment]
No idea why cdecl's example is not actually a valid declaration.

**Comment by matter123** [comment]
There should be support for `int *(*foo)()` and `int &(*foo)()` right?

**Comment by jeff-hykin** [comment]
yeah, both of them compile

**Comment by matter123** [comment]
Ok added support and update test

**Comment by matter123** [comment]
Screenshot from 2019-05-10 16-54-42

**Comment by jeff-hykin** [comment]
So is something like `int (*fcnPtr[])(void);` an array of function pointers?

**Comment by matter123** [comment]
Yep

**Comment by matter123** [comment]
Your dumb ptr/ref allows for `int **&(*foo)();` which isnt valid

**Comment by jeff-hykin** [comment]
👍 working on a fix

**Comment by matter123** [comment]
Also, it looks like the * is no longer tagged.

**Comment by jeff-hykin** [comment]
Not sure which of these is valid according to the spec
Screen Shot 2019-05-10 at 7 15 00 PM

**Comment by jeff-hykin** [comment]
Ah I just realized I left the `[]` in those last two examples

**Comment by matter123** [comment]
yeah because rr value references aren't a thing.
so any number of `*` then up to 2 `&`

**Comment by matter123** [comment]
It should just be
```
optional_type_modifier = maybe(@spaces).then(
            match: zeroOrMoreOf(/\*/).maybe(/&&?/),
            # TODO maybe change see https://github.com/jeff-hykin/cpp-textmate-grammar/pull/135#issuecomment-490727262
            tag_as: "keyword.operator",
        ).maybe(@spaces)
```

Edit: yours is betetr i didn't see it yet.

File: cpp/generate.rb [hunk 1]
```diff
@@ -28,6 +28,16 @@
             match: /,/,
             tag_as: "comma punctuation.separator.delimiter"
         )
+    # TODO maybe change this tag see https://github.com/jeff-hykin/cpp-textmate-grammar/pull/135#issuecomment-490727262
+    # TODO eventually move this outside of the # Utils section
+    ref_deref_definition_pattern = newPattern(
+            match: zeroOrMoreOf(/\*/.maybe(@spaces)),
+            tag_as: "keyword.operator.dereference"
```
Comment (matter123) [inline cpp/generate.rb]: that's a pointer, not a dereference.

Comment (jeff-hykin) [reply cpp/generate.rb]: True. Okay, I guess if were going to make it accurate we should decide on something other than `keyword.operator`. Do you think `storage.modifier.pointer` and `storage.modifier.reference` are accurate?

**Comment by matter123** [comment]
LGTM. I'll update the tests.

**Comment by matter123** [comment]
Is it intentional that the * in `*fncPtr` only has a punctuation scope?

**Comment by matter123** [comment]
This broke tests:
-  `issues/030.cpp`
- `issues/088.cpp`

**Comment by jeff-hykin** [comment]
> Is it intentional that the * in `*fncPtr` only has a punctuation scope?

It was, but its up for debate. I think punctuation (in contrast to operator) is for anything thats telling the compiler where/what something is. Its debatable that it could be part of the `storage.modifier` since its a function pointer.

**Comment by matter123** [comment]
any idea why the test for #30 is broken?

**Comment by jeff-hykin** [comment]
Looking into it now, I'm not immediately sure why though

**Comment by jeff-hykin** [comment]
Oh I forgot about `# TODO this is here because it needs to activate inside of function-pointer parameters. Once function-pointer syntax is implemented, remove it from here` and I think that actually might be the cause

**Comment by matter123** [comment]
Ok, im fixing up test 26.

**Comment by matter123** [comment]
Screenshot from 2019-05-10 18-14-11

**Comment by matter123** [comment]
Anything else that needs to happen before his gets merged in?

**Comment by jeff-hykin** [comment]
I think everything is good 👍

**Pull request merged**